File: app/src/main/java/com/github/premnirmal/ticker/portfolio/StocksAdapter.java
package com.github.premnirmal.ticker.portfolio;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.github.premnirmal.ticker.model.IStocksProvider;
import com.github.premnirmal.ticker.network.Stock;
import com.github.premnirmal.tickerwidget.R;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by premnirmal on 12/21/14.
 */
class StocksAdapter extends RecyclerView.Adapter<StockVH> {

    private final List<Stock> stockList;
    private final OnStockClickListener listener;

    interface OnStockClickListener {
        void onRemoveClick(View view, Stock stock, int position);
        void onClick(Stock stock);
    }

    StocksAdapter(IStocksProvider stocksProvider, OnStockClickListener listener) {
        this.listener = listener;
        stockList = stocksProvider.getStocks() == null
                ? new ArrayList<Stock>()
                : new ArrayList<>(stocksProvider.getStocks());
    }

    boolean remove(final Stock stock) {
        return stockList.remove(stock);
    }

    void refresh(IStocksProvider stocksProvider) {
        stockList.clear();
        stockList.addAll(stocksProvider.getStocks() == null
                ? new ArrayList<Stock>()
                : stocksProvider.getStocks());
        notifyDataSetChanged();
    }

    @Override
    public StockVH onCreateViewHolder(ViewGroup parent, int viewType) {
        final Context context = parent.getContext();
        final View itemView = LayoutInflater.from(context).inflate(R.layout.portfolio_item_view, null);
        return new StockVH(itemView);
    }

    @Override
    public void onBindViewHolder(StockVH holder, int position) {
        holder.update(stockList.get(position), listener);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public int getItemCount() {
        return stockList.size();
    }
}

File: app/src/main/kotlin/com/github/premnirmal/ticker/portfolio/StocksAdapter.kt
package com.github.premnirmal.ticker.portfolio

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import com.github.premnirmal.ticker.network.data.Quote
import com.github.premnirmal.ticker.portfolio.PortfolioVH.PositionVH
import com.github.premnirmal.ticker.portfolio.PortfolioVH.StockVH
import com.github.premnirmal.ticker.portfolio.drag_drop.ItemTouchHelperAdapter
import com.github.premnirmal.ticker.portfolio.drag_drop.OnStartDragListener
import com.github.premnirmal.ticker.widget.WidgetData
import com.github.premnirmal.tickerwidget.databinding.ItemPositionBinding
import com.github.premnirmal.tickerwidget.databinding.ItemStockBinding

/**
 * Created by premnirmal on 2/29/16.
 */
class StocksAdapter constructor(
  private val widgetData: WidgetData,
  private val listener: QuoteClickListener,
  private val dragStartListener: OnStartDragListener
) : ListAdapter<Quote, PortfolioVH>(DiffCallback()), ItemTouchHelperAdapter {

  interface QuoteClickListener {
    fun onClickQuoteOptions(
      view: View,
      quote: Quote,
      position: Int
    ) {}

    fun onOpenQuote(
      view: View,
      quote: Quote,
      position: Int
    )
  }

  companion object {
    const val TYPE_STOCK = 1
    const val TYPE_POSITION = 2
  }

  fun refresh() {
    submitList(widgetData.getQuotesList())
  }

  override fun getItemViewType(position: Int): Int {
    val stock = currentList[position]
    return when {
      stock.hasPositions() -> TYPE_POSITION
      else -> TYPE_STOCK
    }
  }

  override fun onCreateViewHolder(
    parent: ViewGroup,
    viewType: Int
  ): PortfolioVH {
    val context = parent.context
    val portfolioVH: PortfolioVH = if (viewType == TYPE_POSITION) {
      val binding = ItemPositionBinding.inflate(LayoutInflater.from(context), parent, false)
      PositionVH(binding)
    } else {
      val binding = ItemStockBinding.inflate(LayoutInflater.from(context), parent, false)
      StockVH(binding)
    }
    return portfolioVH
  }

  override fun onBindViewHolder(
    holder: PortfolioVH,
    position: Int
  ) {
    holder.update(currentList[position], listener)
    holder.itemView.setOnLongClickListener {
      dragStartListener.onStartDrag(holder)
      true
    }
  }

  override fun getItemId(position: Int): Long = position.toLong()

  override fun onItemMove(
    fromPosition: Int,
    toPosition: Int
  ): Boolean {
    val data = ArrayList(widgetData.getQuotesList())
    data.add(toPosition, data.removeAt(fromPosition))
    val newTickerList = data.mapTo(ArrayList()) { it.symbol }
    widgetData.rearrange(newTickerList)
    notifyItemMoved(fromPosition, toPosition)
    return true
  }

  override fun onItemDismiss(position: Int) {
    dragStartListener.onStopDrag()
  }

  class DiffCallback : DiffUtil.ItemCallback<Quote>() {
    override fun areItemsTheSame(
      oldItem: Quote,
      newItem: Quote
    ): Boolean {
      return oldItem.symbol == newItem.symbol
    }

    override fun areContentsTheSame(
      oldItem: Quote,
      newItem: Quote
    ): Boolean {
      return oldItem == newItem
    }
  }
}
